lighthouse-core/report/html/renderer/details-renderer.js

```diff
@@ -375,49 +406,66 @@ class DetailsRenderer {
 
     const tbodyElem = this._dom.createChildOf(tableElem, 'tbody');
     for (const row of details.items) {
-      const rowElem = this._dom.createChildOf(tbodyElem, 'tr');
+      // Render the main row.
+      const headingAndValuePairs = [];
       for (const heading of headings) {
-        const valueFragment = this._dom.createFragment();
-
-        if (heading.key === null && !heading.subRows) {
-          // eslint-disable-next-line no-console
-          console.warn('A header with a null `key` should define `subRows`.');
+        const value = heading.key && row[heading.key];
+        if (value === undefined || value === null || Array.isArray(value)) {
+          headingAndValuePairs.push(null);
+          continue;
         }
 
-        if (heading.key === null) {
-          const emptyElement = this._dom.createElement('div');
-          emptyElement.innerHTML = '&nbsp;';
-          valueFragment.appendChild(emptyElement);
-        } else {
-          const value = row[heading.key];
-          const valueElement =
-            value !== undefined && !Array.isArray(value) && this._renderTableValue(value, heading);
-          if (valueElement) valueFragment.appendChild(valueElement);
-        }
+        headingAndValuePairs.push({heading, value});
+      }
+
+      tbodyElem.append(this._renderRow(headingAndValuePairs));
+
+      // A single row data value can expand into multiple table rows. These additional table rows
+      // are called sub-rows.
+
+      const subRowHeadings = headings.map(heading => {
+        if (!heading.subRows) return null;
+        return {
+          key: heading.subRows.key,
+          valueType: heading.subRows.valueType || heading.valueType,
+          granularity: heading.subRows.granularity || heading.granularity,
+          displayUnit: heading.subRows.displayUnit || heading.displayUnit,
+          label: '',
+        };
+      });
 
-        if (heading.subRows) {
-          const subRowsHeading = {
-            key: heading.subRows.key,
-            valueType: heading.subRows.valueType || heading.valueType,
-            granularity: heading.subRows.granularity || heading.granularity,
-            displayUnit: heading.subRows.displayUnit || heading.displayUnit,
-            label: '',
-          };
-          const values = row[subRowsHeading.key];
-          if (Array.isArray(values)) {
-            const subRowsElement = this._renderSubRows(values, subRowsHeading);
-            valueFragment.appendChild(subRowsElement);
+      if (!subRowHeadings.some(Boolean)) continue;
+
+      // All sub-row data arrays should be the same length, but just in case they are not,
+      // determine the longest data array and fill the missing values with an
+      // null pair (which creates an empty cell).
```

Comment:
adding nulls to create empty cells seems very weird to me. 
its to maintain alignment with the parent headings right? but dont the parent headings already know what their subheadings are? can we use that relationship?

Reply:
background color only colors the width of the table if there is an element there.

Reply:
Paul refactor to remove the nulls here: https://github.com/GoogleChrome/lighthouse/compare/subrow-layout...subrow-layout-by-col

I think most of the complexity in your and my impl. comes from trying to pass the values to this `_renderRow` function. How about instead the fn uses a callback? I think it'd look like this:

```
_renderRow(headings, heading => row && row[heading.key])

for i in numSubRow:
  const subRowEl = _renderRow(headings, heading => row && Array.isArray(row[heading.key]) && row[heading.key][i])
  ...
```

Reply:
yeah thats a nice solution. :)

Reply:
Peanut gallery: this seems like it's getting pretty involved for a static, self-generated table json of known width and height being turned into a `<table>` of known width and height. There's some table cruft we have to live with for legacy support, but subrows are brand new :)

Since (I don't believe) subrows are turned on anywhere yet (and we haven't shipped 6.0 yet regardless), maybe knowing more of the use case now gives a chance to step back and massage the data structure a bit to make dealing with them simpler? (#&#x2060;5 on the classic https://users.ece.utexas.edu/~adnan/pike.html)